llvm/lib/Transforms/Utils/CancelCrossCUDebugInfo.cpp

```diff
@@ -0,0 +1,135 @@
+//===- CancelCrossCUDebugInfo.cpp - Enable cross cu canceling--------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/Transforms/Utils/CancelCrossCUDebugInfo.h"
+#include "llvm/IR/DIBuilder.h"
+#include "llvm/IR/DebugInfoMetadata.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/InitializePasses.h"
+#include "llvm/Pass.h"
+#include "llvm/Transforms/Utils.h"
+#include "llvm/Transforms/Utils/ValueMapper.h"
+
+using namespace llvm;
+
+bool isFunctionSkipped(Function &F) {
+  return F.isDeclaration() || !F.hasExactDefinition();
+}
+
+PreservedAnalyses CancelCrossCUDebugInfo::run(Module &M,
+                                              ModuleAnalysisManager &MA) {
+
+  for (Function &F : M) {
+    if (isFunctionSkipped(F))
+      continue;
+
+    Module &M = *F.getParent();
+    DICompileUnit *cu = *M.debug_compile_units_begin();
+
+    DIBuilder DIB(M, true, cu);
+    LLVMContext &Ctx = M.getContext();
+
+    std::map<StringRef, DISubprogram *> FunctionNameMap;
+
+    auto OrigSP = F.getSubprogram();
+
+    ValueToValueMapTy VMap;
+
+    auto mapMDtoMD = [&VMap](MDNode *Old, MDNode *New) {
+      // Avoid clobbering an existing mapping.
+      (void)VMap.MD().try_emplace(Old, New);
+    };
+    DISubprogram *SPInline = nullptr;
+    for (BasicBlock &BB : F) {
+      for (Instruction &I : BB) {
+        auto DebugLoc = I.getDebugLoc();
+
+        auto InlineLoc = DebugLoc.getInlinedAt();
+
+        if (InlineLoc) {
+          auto Scope = DebugLoc.getScope();
+          if (auto *T = dyn_cast<DISubprogram>(Scope)) {
+            if (T->getFilename() != OrigSP->getFilename()) {
+
+              if (FunctionNameMap.find(T->getFilename()) ==
+                  FunctionNameMap.end()) {
+
+                SPInline = DIB.createFunction(
+                    OrigSP->getFile(), T->getName(), T->getName(),
```

Comment:
AFAICT, this is not the containing DICompileUnit; that is in the DIB.  It is wrong to change the filename (either scope or file) and will break debugging.  We only want to replicate any content (including the DIFile) into the other CompileUnit.